### ISSUE TITLE
Is it possible for Copy to loop forever?

### ISSUE THREAD

**Issue by EFanZh** [tokio-rs/tokio-io#56]
Here is some code from https://github.com/tokio-rs/tokio-io/blob/master/src/copy.rs#L73-L78:

```rust
while self.pos < self.cap {
    let writer = self.writer.as_mut().unwrap();
    let i = try_nb!(writer.write(&self.buf[self.pos..self.cap]));
    self.pos += i;
    self.amt += i as u64;
}
```

Is it possible for `try_nb!(writer.write(&self.buf[self.pos..self.cap]))` to return `0`? If so, wouldn’t it cause the code to loop forever?

**Comment by alexcrichton** [comment]
@EFanZh ah yes indeed! I think that this should explicitly handle the 0-write case and return an error. Do you want to send a PR?

**Comment by EFanZh** [comment]
@alexcrichton Sure. A pull request is submitted: #57.

**Comment by alexcrichton** [comment]
Fixed in #57
